### PR TITLE
9102 zfs should be able to initialize storage devices

### PR DESCRIPTION
**Pull request by grwilson**
Reviewed by: John Wren Kennedy <john.kennedy@delphix.com>
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>
Reviewed by: Prakash Surya <prakash.surya@delphix.com>

PROBLEM
========
The first access to a block incurs a performance penalty on some platforms
(e.g. AWS's EBS, VMware VMDKs). Therefore we recommend that volumes are "thick
provisioned", where supported by the platform (VMware). This can create a large
delay in getting a new virtual machines up and running (or adding storage to
an existing Engine). If the thick provision step is omitted, write  performance
will be suboptimal until all blocks on the LUN have been written.

SOLUTION
=========
This feature introduces a way to 'initialize' the disks at install or in the
background to make sure we don't incur this first read penalty.

When an entire LUN is added to ZFS, we make all space available immediately,
and allow ZFS to find unallocated space and zero it out. This works with
concurrent writes to arbitrary offsets, ensuring that we don't zero out
something that has been (or is in the middle of being) written. This scheme
can also be applied to existing pools (affecting only free regions on the
vdev). Detailed design:
        - new subcommand:zpool initialize [-cs] <pool> [<vdev> ...]
                - start, suspend, or cancel initialization
        - Creates new open-context thread for each vdev
        - Thread iterates through all metaslabs in this vdev
        - Each metaslab:
                - select a metaslab
                - load the metaslab
                - mark the metaslab as being zeroed
                - walk all free ranges within that metaslab and translate them
                  to ranges on the leaf vdev
                - issue a "zeroing" I/O on the leaf vdev that corresponds to a
                  free range on the metaslab we're working on
                - continue until all free ranges for this metaslab have been
                  "zeroed"
                - reset/unmark the metaslab being zeroed
                - if more metaslabs exist, then repeat above tasks.
                - if no more metaslabs, then we're done.

        - progress for the initialization is stored on-disk in the vdev’s leaf
          zap object. The following information is stored:
                - the last offset that has been initialized
                - the state of the initialization process (i.e. active,
                  suspended, or canceled)
                - the start time for the initialization

        - progress is reported via the zpool status command and shows
          information for each of the videos that are initializing

### CODE REVIEW

**Comment by dcarosone** [comment]
Why not just dd a file to a scratch dataset (with no compression, copies=3,
a suitable quota to prevent filling, and whatever other properties you
think might be relevant).

The use case is legitimate, but the need for more code to address it seems
dubious.

**Comment by grwilson** [comment]
You could go through and `dd` to every disk before adding it to the pool and that is the common practice used by many. The problem that this is trying to solve is to be able to avoid the delay of having to wait for every disk to be pre-warmed before using it. Also, people often forget to "warm" a disk before adding it to a pool and then find themselves stuck with the performance penalty. This command allows you to "warm" the disk at anytime. It also means that you could "warm" the disk while you immediately start using it. During that time there would be an I/O penalty since you're obviously doing more I/O than just writes but it allows those operations to run concurrently. Furthermore, you could extend this command to implement some form of secure delete where you want to write a pattern over any free space. Or even manually TRIM the data by issuing TRIM commands on any free region of the disk. This functionality has a very specific use case for us at Delphix but we felt that others might be interested in something similar (especially since people today are already using `dd` for this purpose).

**Comment by ahrens** [comment]
I think that @dcarosone is asking about creating a file in the pool to force zfs to write to most of the pool (not dd-ing to the raw disk).  That could work but the `zpool initialize` approach has a few advantages:

 - You don't need to figure out how big to make the file.  The file size can't be determined in advance (because you'll be filling the pool with "real" data while this is happening - that's the whole point), so you'll have to monitor the pool fullness and adjust the amount of data to write.  Since this would be a userland process, it would be more error prone and less portable.
 - You'll be sure that all the parts of the LUNs are initialized.  With the file approach, you'll have to leave some free space, which won't be initialized.
 - When adding a device to an existing pool, you'll only initialize the new device.  With the file approach, ZFS would also (re-)write to the free space in the existing devices.

**Comment by ahrens** [comment]
Igor K has also reviewed and tested this.

**Comment by prakashsurya** [comment]
@grwilson can you rebase this with the last master code when you get a chance?

**Comment by dlethe** [comment]
I don't see the practicality of this in most cases.  The logic insures that a pool will do unnecessary write I/Os on parts of the pool that certainly will never get written to ... ever.   

This will also not insure that all physical blocks even get initialized in the first place. (you must not skip over all areas of disk used for housekeeping, dumps, swap, metadata, etc ...)   Depending on the configuration of the disks, expander, and backplane you could start saturating devices and then you'll *really* get a performance hit much worse than if you did nothing at all.

Initialization requires 100% of them in order to calculate ECC Data. The penalty is approximately 50% hit on first write, and reading a block that has not been initialized can be much, much worse as it can result in ECC recovery.  So if you even do this, you had better get 100% of all physical blocks on all disks, and leave nothing to chance.  What if you have to resilver?? 

 So bottom line, unless EVERY LBA of the disk drive that could ever be read from, even slices not used for ZFS, are initialized, then perhaps this entire project should be suspended until performance is modeled on uninitialized storage before doing any more development.

**Comment by ahrens** [comment]
@dlethe I think you're thinking about a different use case than we are.  We want to avoid the first-write penalty on virtualized storage (e.g. ESX VMFS, AWS EBS, etc).  But perhaps your use case would also benefit from initialization!  Let me address a few of your specific points:

> The logic insures that a pool will do unnecessary write I/Os on parts of the pool that certainly will never get written to ... ever.

How do you know that?  It's highly likely that all parts of the pool will be eventually written to, even if the pool never gets close to being "full", due to copy-on-write.

> This will also not insure that all physical blocks even get initialized in the first place. (you must not skip over all areas of disk used for housekeeping, dumps, swap, metadata, etc ...)

Swap [zvols] and [ZFS] metadata are normal parts of the pool so there's nothing special there - since they've been written, there's no need to write them again to initialize.  Reservations for zvols are just an accounting trick, so that unallocated space will be initialized.

Dump [zvols] are a special case and I think you're right that they won't be initialized, since we pre-allocate them and then write in place.  You could get around this by doing the initialize before creating the dump zvol.  Or (as in our case) by having different pools for root and production data (and not worrying too much about performance of the root pool, and dump in particular).

I'm not sure what you mean by "housekeeping".

> Initialization requires 100% of them in order to calculate ECC Data. The penalty is approximately 50% hit on first write, and reading a block that has not been initialized can be much, much worse as it can result in ECC recovery. So if you even do this, you had better get 100% of all physical blocks on all disks, and leave nothing to chance. What if you have to resilver??

This is definitely a different use case than we had in mind.  At what granularity is the ECC generated?  Surely it isn't one ECC value for the entire disk?  It seems like the ECC must be per "chunk", and if that "chunk" has been entirely written then you get this benefit.  So if we can get 99.99% of the chunks, that seems worthwhile, since the vast majority of future writes will get the benefit of initialization.

> So bottom line, unless EVERY LBA of the disk drive that could ever be read from, even slices not used for ZFS, are initialized, 

Can you expand on your thought process as to why every single block needs to be written?  Why doesn't initializing 99% of the blocks result in 99% of the benefit?

> then perhaps this entire project should be suspended until performance is modeled on uninitialized storage

We did model performance on uninitialized [EBS and VMFS] storage.  I don't see that mentioned here so I'll try to find the results.

> before doing any more development.

We aren't doing any more development either way -  this project is complete and ready to integrate.

**Comment by dlethe** [comment]
Yes, @ahrens - I was incorrectly thinking of use case associated with initializing physical blocks when disk is either TCG/SED drive, or God-forbid, using some sort of RAID device in where there is already some sort of logical/physical mapping.  

* I did mention some parts of disk never getting written to, certainly there must be parts, especially if the pool has a heck of a lot of free space.   The more free space you have, the more writes that get done needlessly, which means unnecessary write overhead.  

* In cases of the initialization required with TCG/SED drives, then each physical block of data (whatever the internal size is) has approximately a 50% penalty on first write or read per the SEAGATE and HGST product manuals.  That is because once the bands are configured and the 256-bit "password" is set, then a full write pass is required to compute ECC for the bits on the disk that are not addressable via the vanilla READ/WRITE CDBs

* To get the total number of bits you need to look at the specs of a drive where it talks about how error rate ie, 1 x 10E15 or whatever. But these numbers are nothing you have to worry about as those ECC bits are maintained inside the SSD or HDD firmware.   

* There are corner cases where physical data that was never written to using this mechanism would end up being read.  From a SCSI engineer perspective, read-ahead,pre-fetch settings in mode pages could cause disks to read data that never had ECC data written.

* In the TCG/SED initialization use case, this mechanism as described is unsuitable, because still a user could always just do a dd at the /dev/rdsk level and read uninitialized blocks.   But as you said, the use case I am thinking of is different then the one you are designing for, so all the inherent problems you would get with a partially initialized target are moot.
 
So basically outside of making sure the dump and potentially looking at whether some application is doing raw I/O at physical device then just warning people about appropriate use cases should be sufficient.

(Well, maybe just consider a thought experiment with a freshly built RAIDZ2 pool using 8TB disks for a large data repository that only has a few filesystems to start)

**Comment by ahrens** [comment]
@prakashsurya internal test link: http://platform.jenkins.delphix.com/job/devops-gate/job/master/job/zfs-precommit/3168/flowGraphTable/

**Comment by prakashsurya** [comment]
@grwilson @ahrens Can either of you rebase this onto the latest master? The RTI has been approved, but there's a small merge conflict. I'm happy to push once it applies cleanly.

**Comment by ahrens** [comment]
@prakashsurya rebased.

**Comment by ikozhukhov** [comment]
what is ti status of this PR?

**Comment by citrus-it** [comment]
I know it's already integrated but the original description was about `zeroing` out the unused blocks and the implementation actually writes `0xdeadbeef` which makes it less useful for me in the case where I want to zero the blocks prior to doing a hole-punch on a sparse VM disk (and I appreciate that is the opposite to the original intention of this change).

Any reason why it is writing deadbeef instead of zeros? I see that the tests explicitly set the pattern to deadbeef for the duration of the test.

**Comment by grwilson** [comment]
@citrus-it in the past we've seen storage arrays that have zeroed out blocks behind the scenes leading to ZFS reported checksums. We wanted to have a default value that would not be confused with that type of corruption. We did make this a global parameter (`zfs_initialize_value`) so that it could be changed via mdb or /etc/system (or other mechanisms). Maybe a future enhancement would be to make this a true parameter to the pool.
